main/Main.hs

```diff
@@ -94,138 +87,167 @@ main = do
+        runSettings serverSettings postgrestApplication
 
-  -- reload schema cache + config on NOTIFY
-  when dbChannelEnabled $
-    listener dbUri dbChannel pool refConf refDbStructure mvarConnectionStatus connWorker fullConfigReReader
+setBuffering :: IO ()
+setBuffering = do
+  -- LineBuffering: the entire output buffer is flushed whenever a newline is
+  -- output, the buffer overflows, a hFlush is issued or the handle is closed
+  hSetBuffering stdout LineBuffering
+  hSetBuffering stdin LineBuffering
+  -- NoBuffering: output is written immediately and never stored in the buffer
+  hSetBuffering stderr NoBuffering
 
-  -- ask for the OS time at most once per second
-  getTime <- mkAutoUpdate defaultUpdateSettings {updateAction = getCurrentTime}
 
-  let postgrestApplication =
-        postgrest
-          logLevel
-          refConf
-          refDbStructure
-          pool
-          getTime
-          connWorker
+-- SIGNALS
 
 #ifndef mingw32_HOST_OS
-  -- run the postgrest application with user defined socket. Only for UNIX systems.
-  whenJust maybeSocketAddr $
-    runAppInSocket serverSettings postgrestApplication socketFileMode
+data SignalHandlers = SignalHandlers
+  { fullConfigReReader :: IO ()
+  , connWorker         :: IO ()
+  , releasePool        :: IO ()
+  }
+
+-- | Set signal handlers, only for systems with signals
+installHandlers :: SignalHandlers -> IO ()
+installHandlers SignalHandlers{..} = do
+  -- Releases the connection pool whenever the program is terminated,
+  -- see https://github.com/PostgREST/postgrest/issues/268
+  install sigINT releasePool
+  install sigTERM releasePool
+
+  -- The SIGUSR1 signal updates the internal 'DbStructure' by running
+  -- 'connectionWorker' exactly as before.
+  install sigUSR1 connWorker
+
+  -- Re-read the config on SIGUSR2
+  install sigUSR2 fullConfigReReader
+  where
+    install signal handler = void $ installHandler signal (Catch handler) Nothing
 #endif
 
-  -- run the postgrest application
-  whenNothing maybeSocketAddr $ do
-    putStrLn $ ("Listening on port " :: Text) <> show port
-    runSettings serverSettings postgrestApplication
 
-readEnvironment :: IO Environment
-readEnvironment = getEnvironment <&> pgrst
+-- APP STATE
+
+
+data AppState = AppState
+  { statePool             :: P.Pool
+  -- ^ Connection pool, either a 'Connection' or a 'ConnectionError'
+  , stateConnectionStatus :: MVar ConnectionStatus
+  -- ^ Used to sync the listener(NOTIFY reload) with the connectionWorker. No
+  -- connection for the listener at first. Only used if dbChannelEnabled=true.
+  , stateDbStructure      :: IORef (Maybe DbStructure)
+  -- ^ No schema cache at the start. Will be filled in by the connectionWorker
+  , stateIsWorkerOn       :: IORef Bool
+  -- ^ Helper ref to make sure just one connectionWorker can run at a time
+  , stateConf             :: IORef AppConfig
+  -- ^ Config that can change at runtime
+  , stateGetTime          :: IO UTCTime
+  }
```

Comment:
Cool. Is this essentially what I had down as `Context` in my graph? i.e. - do you plan to move this out and nest config/schema etc. under `AppState`? Or is this just an internal helper here?

Reply:
Yes, will be split out! As a fully opaque type even I think